### PR TITLE
Node labelling UX

### PR DESCRIPTION
**Pull request by tiagolobocastro**
❯ rest-plugin label node io-engine-1 openebs.io/performance=good
Node io-engine-1 labelled successfully. Current labels: {"openebs.io/performance": "good"}

❯ rest-plugin label node io-engine-1 openebs.io/performance=good
Node io-engine-1 not labelled as the label key already exists

❯ rest-plugin label node io-engine-1 openebs.io/performance=bad
Node io-engine-1 not labelled as the label already exists with different value

❯ rest-plugin label node io-engine-1 openebs.io/performance/also=good
Invalid node label: / is only allowed once

❯ rest-plugin label node io-engine-1 openebs.io/performance/=good
Invalid node label: / can only be used in the middle of the key

❯ rest-plugin label node io-engine-1 openebs.io/performance
Invalid node label: At least one label value is required

❯ rest-plugin label node io-engine-1 openebs.io/performance-
Node io-engine-1 labelled successfully. Current labels: {}

❯ rest-plugin label node io-engine-1 openebs.io/performances-
Node io-engine-1 not unlabelled as it did not contain the labe

---

    refactor: apply python linter suggestions
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    ci: check if python linter modified files
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    ci: don't install pre-commit on CI
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    refactor(node-label/api): doc the cli and openapi with key val
    
    Expands the parsing of label on the cli and returns different error messages.
    Splits the label into key and value.
    todo: is this the right approach here, where should we do the validation?
    kubectl seems to some validation, is there anything on the apiserver too?
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    refactor(node-label): use specific types for store operations
    
    PStor operations should use a specific type to allow for better expansion.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    docs(node-label/rest-plugin): improve ux for the command line
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    build(openapi-generator): upgrade to v6.4.0
    
    Namely fixes a bug related to path parsing with '='
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

### CODE REVIEW

File: control-plane/plugin/src/resources/error.rs [hunk 1]
```diff
@@ -24,13 +24,13 @@ pub enum Error {
         source: openapi::tower::client::Error<openapi::models::RestJsonError>,
     },
     /// Error when node label request fails.
-    #[snafu(display("Failed to get node {id}. Error {source}"))]
-    NodeLabelError {
+    #[snafu(display("Failed label node {id}. Error {source}"))]
```
Comment (Abhinandan-Purkait) [inline control-plane/plugin/src/resources/error.rs]: `Failed label` --> `Failed to label` ?

File: control-plane/plugin/src/resources/mod.rs [hunk 2]
```diff
@@ -104,10 +104,22 @@ pub enum DrainResources {
 /// The types of resources that support the 'label' operation.
 #[derive(clap::Subcommand, Debug)]
 pub enum LabelResources {
-    /// Label node with the given ID.
+    /// Adds or removes a label from the specified node.
```
Comment (Abhinandan-Purkait) [inline control-plane/plugin/src/resources/mod.rs]: `Adds or removes a label from the specified node.` --> `Adds or removes a label to or from the specified node.`

File: control-plane/plugin/src/resources/mod.rs [hunk 3]
```diff
@@ -104,10 +104,22 @@ pub enum DrainResources {
...
     Node {
+        /// The id of the node to label.
         id: NodeId,
+        /// The label to be added or removed from the node.
+        /// To add a label, specify the following format:
+        /// ${key}=${value}
+        /// To remove a label, specify the following format:
+        /// ${key}-
+        /// A label key and value must begin with a letter or number, and may contain letters,
```
Comment (Abhinandan-Purkait) [inline control-plane/plugin/src/resources/mod.rs:116]: So key beginning with a number is a valid key?

File: control-plane/plugin/src/resources/node.rs [hunk 4]
```diff
@@ -555,6 +555,49 @@ impl Drain for Node {
...
+    }
+    if key.starts_with('/') || key.ends_with('/') {
+        return Err(Error::NodeLabelFormat {
+            error: "/ can only be used in the middle of the key".to_string(),
+        });
+    }
+
+    if key.chars().filter(|c| c == &'/').count() > 1 {
+        return Err(Error::NodeLabelFormat {
+            error: "/ is only allowed once".to_string(),
```
Comment (Abhinandan-Purkait) [inline control-plane/plugin/src/resources/node.rs]: "/ is only allowed once"
```suggestion
            error: "/ is allowed only once".to_string(),
```
Does this sound better? Maybe the above one as well?

**Review by Abhinandan-Purkait — APPROVED** [review]
LGTM

File: control-plane/plugin/src/resources/mod.rs (hunk 3 above)
Comment (tiagolobocastro) [reply control-plane/plugin/src/resources/mod.rs:116]: Yes it is

File: control-plane/plugin/src/resources/node.rs (hunk 4 above)
Comment (tiagolobocastro) [reply control-plane/plugin/src/resources/node.rs]: Both are the same to me but I can change to the latter if you prefer

Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/resources/node.rs]: No, upto you, I think it's fine either way.

**Review by sinhaashish — APPROVED** [review]
LGTM

File: control-plane/plugin/src/resources/node.rs (hunk 4 above)
Comment (tiagolobocastro) [reply control-plane/plugin/src/resources/node.rs]: I tweaked it a bit as I moved the errors into an enum

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**